docs/design_notes/module_lifecycle.rst

```diff
@@ -0,0 +1,378 @@
+
+Construction/initialization
+=============================
+
+Now we want to construct and use the ``MLP`` Module:
+
+
+.. testcode::
+
+  mlp = MLP(hidden_size=5, out_size=3)
+  x = jax.numpy.ones((1, 2))
+  variables = mlp.init(random.PRNGKey(0), x)
+  y = mlp.apply(variables, x)
+
+
+First, we construct an instance of ``MLP`` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The ``MLP`` constructor does not actually create variables or any internal state whatsoever. It's best to think of it as a specification or template of the Module that contains functionality but no data.
+
+Let's first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling ``init`` is just a special case of ``apply`` and we could also code it as:
+
+
+.. testcode:
+
+  # equivalent to: variables = mlp.init(random.PRNGKey(0), x)
+  _, variables = mlp.apply({}, x, rngs={"params": random.PRNGKey(0)}, mutable=True)
+
+
+Thus, ``init`` is nothing more than a special case of ``apply`` where:
+
+#. We call a Module without any initial variables (an empty dict).
+#. A PRNG generator named “params” is always passed for randomly initializing parameters (using the parameter initialization function).
+#. All variable collections are set to mutable (``mutable=True``). When a collection is mutable, existing variables can be updated and new variables can be created. Thus, inside ``init`` variables can be initialized in any variable collection and they are all added to the returned variable dictionary.
+
+Lifecycle
+=============
+
+
+Now that we understand ``init`` is just a special case of ``apply``, let's analyze what ``.apply(...)`` does in more detail. In fact, most of the complexity of Modules resides in the ``apply`` method. The "Module lifecycle" consists of constructing and ``apply``-ing a Module. We can summarize the Module lifecycle as follows:
+
+
+#. We construct `mlp = MLP(hidden_size=5, out_size=3)`, such that `mlp.hidden_size=5` and `mlp_out_size=3`.
+
+#. Then, call ``mlp.apply`` which performs the following:
+
+   #. Makes a clone of ``mlp``, let's call it `mlp_copy`.
+
+   #. Calls `mlp_copy.setup()`.
+
+   #. Returns the output of ``mlp_copy.__call__()`` and optionally the variable collections that were specified as mutable using the keyword argument ``mutable=``.
+
+
+You might be surprised that the lifecycle includes cloning the Module instance. You might be surprised that the lifecycle includes cloning the Module instance. In short, this is done to ensure that `apply` can be treated as a pure function (i.e., if you pass the same arguments in, it will return the same outputs). This will be explained in more detail later in the  :ref:`Top-level Modules` section.
```

Comment:
Nit: repeating sentences.

```suggestion
Notice that the lifecycle includes cloning the Module instance. This is done to ensure that `apply` can be treated as a pure function (i.e., if you pass the same arguments in, it will return the same outputs). You will learn about this in more detail later in the  :ref:`Top-level Modules` section.
```

(+ some minor refactoring)